File: src/calcio_manager/models/team.py
"""Team data models."""

from uuid import UUID, uuid4

from pydantic import BaseModel, Field

from calcio_manager.models.enums import PlayerRole, TacticStyle
from calcio_manager.models.player import Player


class Formation(BaseModel):
    """A team formation for C7 (excludes goalkeeper)."""

    name: str  # e.g., "2-3-1"
    defenders: int = Field(ge=0, le=6)
    midfielders: int = Field(ge=0, le=6)
    forwards: int = Field(ge=0, le=6)

    def model_post_init(self, __context: object) -> None:
        """Validate that formation adds up to 6 outfield players."""
        total = self.defenders + self.midfielders + self.forwards
        if total != 6:
            raise ValueError(
                f"Formation {self.name} has {total} outfield players, expected 6"
            )

    @classmethod
    def default_formations(cls) -> list["Formation"]:
        """Return the standard C7 formations."""
        return [
            cls(name="2-3-1", defenders=2, midfielders=3, forwards=1),
            cls(name="2-2-2", defenders=2, midfielders=2, forwards=2),
            cls(name="3-2-1", defenders=3, midfielders=2, forwards=1),
            cls(name="3-1-2", defenders=3, midfielders=1, forwards=2),
            cls(name="1-3-2", defenders=1, midfielders=3, forwards=2),
            cls(name="2-1-3", defenders=2, midfielders=1, forwards=3),
        ]


class Lineup(BaseModel):
    """Starting lineup and substitutes for a match."""

    starters: dict[str, UUID] = Field(
        default_factory=dict,
        description="Position slot -> Player ID mapping",
    )
    substitutes: list[UUID] = Field(default_factory=list)
    formation: Formation = Formation(name="2-3-1", defenders=2, midfielders=3, forwards=1)

    @property
    def starter_count(self) -> int:
        """Number of starting players."""
        return len(self.starters)


class TeamFinances(BaseModel):
    """Financial state of a team."""

    budget: int = Field(default=2000, description="Available budget in euros")
    registration_fees_paid: int = 0
    sponsor_income: int = 0
    tournament_prizes: int = 0
    expenses: int = 0

    @property
    def balance(self) -> int:
        """Current balance."""
        return (
            self.budget
            + self.sponsor_income
            + self.tournament_prizes
            - self.registration_fees_paid
            - self.expenses
        )


class Team(BaseModel):
    """A football team in the game."""

    id: UUID = Field(default_factory=uuid4)
    name: str
    city: str
    province: str = "Varese"
    colors: tuple[str, str] = ("bianco", "nero")
    reputation: int = Field(ge=1, le=100, default=50)
    squad: list[Player] = Field(default_factory=list)
    formation: Formation = Formation(name="2-3-1", defenders=2, midfielders=3, forwards=1)
    tactic: TacticStyle = TacticStyle.BALANCED
    finances: TeamFinances = TeamFinances()
    is_human: bool = False

    @property
    def available_players(self) -> list[Player]:
        """Return players available for selection (not injured/suspended)."""
        return [p for p in self.squad if p.is_available]

    @property
    def goalkeepers(self) -> list[Player]:
        """Return all goalkeepers in the squad."""
        return [p for p in self.squad if p.role == PlayerRole.GK]

    @property
    def outfield_players(self) -> list[Player]:
        """Return all non-goalkeeper players."""
        return [p for p in self.squad if p.role != PlayerRole.GK]

    @property
    def can_play(self) -> bool:
        """Check if the team has enough available players (CSI minimum: 4)."""
        return len(self.available_players) >= 4

    @property
    def squad_average_overall(self) -> float:
        """Average overall rating of the squad."""
        if not self.squad:
            return 0.0
        return sum(p.overall for p in self.squad) / len(self.squad)

File: src-tauri/src/domain/models/team.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::enums::{PlayerRole, TacticStyle};
use super::player::Player;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Formation {
    pub name: String,
    pub defenders: u8,
    pub midfielders: u8,
    pub forwards: u8,
}

impl Formation {
    pub fn new(
        name: impl Into<String>,
        defenders: u8,
        midfielders: u8,
        forwards: u8,
    ) -> Result<Self, String> {
        let formation = Self {
            name: name.into(),
            defenders,
            midfielders,
            forwards,
        };
        if formation.defenders + formation.midfielders + formation.forwards != 6 {
            return Err(format!(
                "Formation {} must contain 6 outfield players",
                formation.name
            ));
        }
        Ok(formation)
    }

    pub fn default_c7() -> Self {
        Self {
            name: "2-3-1".to_string(),
            defenders: 2,
            midfielders: 3,
            forwards: 1,
        }
    }

    pub fn default_formations() -> Vec<Self> {
        vec![
            Self {
                name: "2-3-1".to_string(),
                defenders: 2,
                midfielders: 3,
                forwards: 1,
            },
            Self {
                name: "2-2-2".to_string(),
                defenders: 2,
                midfielders: 2,
                forwards: 2,
            },
            Self {
                name: "3-2-1".to_string(),
                defenders: 3,
                midfielders: 2,
                forwards: 1,
            },
            Self {
                name: "3-1-2".to_string(),
                defenders: 3,
                midfielders: 1,
                forwards: 2,
            },
            Self {
                name: "1-3-2".to_string(),
                defenders: 1,
                midfielders: 3,
                forwards: 2,
            },
            Self {
                name: "2-1-3".to_string(),
                defenders: 2,
                midfielders: 1,
                forwards: 3,
            },
        ]
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Lineup {
    pub starters: std::collections::HashMap<String, Uuid>,
    pub substitutes: Vec<Uuid>,
    pub formation: Formation,
}

impl Default for Lineup {
    fn default() -> Self {
        Self {
            starters: Default::default(),
            substitutes: vec![],
            formation: Formation::default_c7(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamFinances {
    pub budget: i32,
    pub registration_fees_paid: i32,
    pub sponsor_income: i32,
    pub tournament_prizes: i32,
    pub expenses: i32,
}

impl Default for TeamFinances {
    fn default() -> Self {
        Self {
            budget: 2000,
            registration_fees_paid: 0,
            sponsor_income: 0,
            tournament_prizes: 0,
            expenses: 0,
        }
    }
}

impl TeamFinances {
    pub fn balance(&self) -> i32 {
        self.budget + self.sponsor_income + self.tournament_prizes
            - self.registration_fees_paid
            - self.expenses
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Team {
    pub id: Uuid,
    pub name: String,
    pub city: String,
    pub province: String,
    pub stadium_name: String,
    pub colors: (String, String),
    pub reputation: u8,
    pub squad: Vec<Player>,
    pub formation: Formation,
    pub tactic: TacticStyle,
    pub finances: TeamFinances,
    pub is_human: bool,
}

impl Team {
    pub fn available_players(&self) -> Vec<&Player> {
        self.squad
            .iter()
            .filter(|player| player.is_available())
            .collect()
    }

    pub fn can_play(&self) -> bool {
        self.available_players().len() >= 4
    }

    pub fn squad_average_overall(&self) -> f32 {
        if self.squad.is_empty() {
            return 0.0;
        }
        self.squad.iter().map(|p| p.overall as f32).sum::<f32>() / self.squad.len() as f32
    }

    pub fn role_count(&self, role: PlayerRole) -> usize {
        self.squad
            .iter()
            .filter(|player| player.role == role)
            .count()
    }
}
